File: src/test/java/dev/nytweetdeck/timeline/PostTextUrlNormalizerTest.java
package dev.nytweetdeck.timeline;

import static dev.nytweetdeck.timeline.PostTextUrlNormalizer.Kind.ARTICLE;
import static dev.nytweetdeck.timeline.PostTextUrlNormalizer.Kind.LINK;
import static dev.nytweetdeck.timeline.PostTextUrlNormalizer.Kind.MEDIA;
import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import org.junit.jupiter.api.Test;

class PostTextUrlNormalizerTest {

    @Test
    void expandsOrdinaryLinksAndPrefersTheUnwoundDestination() {
        var entities = List.of(new PostTextUrlNormalizer.UrlEntity(
                "https://t.co/docs", "https://example.com/short", "https://example.com/original?a=1", LINK));

        assertThat(PostTextUrlNormalizer.normalize("資料 https://t.co/docs", entities))
                .isEqualTo("資料 https://example.com/original?a=1");
    }

    @Test
    void removesMediaAndArticleRedirectsWithoutTouchingLookalikeHosts() {
        var entities = List.of(
                new PostTextUrlNormalizer.UrlEntity("https://t.co/media", null, null, MEDIA),
                new PostTextUrlNormalizer.UrlEntity("https://t.co/article", null, null, ARTICLE),
                new PostTextUrlNormalizer.UrlEntity(
                        "https://not-t.co/link", "https://example.com/original", null, LINK));

        assertThat(PostTextUrlNormalizer.normalize(
                        "本文 https://t.co/media https://t.co/article https://not-t.co/link", entities))
                .isEqualTo("本文 https://not-t.co/link");
    }

    @Test
    void keepsTheShortUrlWhenTheDestinationIsMissingOrUnsafe() {
        var entities = List.of(new PostTextUrlNormalizer.UrlEntity(
                "https://t.co/docs", "javascript:alert(1)", null, LINK));

        assertThat(PostTextUrlNormalizer.normalize("https://t.co/docs", entities))
                .isEqualTo("https://t.co/docs");
    }
}

File: android/app/src/test/java/dev/nytweetdeck/android/xapi/PostTextUrlNormalizerTest.kt
package dev.nytweetdeck.android.xapi

import org.junit.Assert.assertEquals
import org.junit.Test

class PostTextUrlNormalizerTest {
    @Test
    fun expandsOrdinaryLinksAndPrefersTheUnwoundDestination() {
        val entities = listOf(
            PostUrlEntity(
                shortUrl = "https://t.co/docs",
                expandedUrl = "https://example.com/short",
                unwoundUrl = "https://example.com/original?a=1",
                kind = PostUrlEntityKind.LINK,
            ),
        )

        assertEquals(
            "資料 https://example.com/original?a=1",
            PostTextUrlNormalizer.normalize("資料 https://t.co/docs", entities),
        )
    }

    @Test
    fun removesMediaAndArticleRedirectsWithoutTouchingLookalikeHosts() {
        val entities = listOf(
            PostUrlEntity("https://t.co/media", null, null, PostUrlEntityKind.MEDIA),
            PostUrlEntity("https://t.co/article", null, null, PostUrlEntityKind.ARTICLE),
            PostUrlEntity(
                "https://not-t.co/link",
                "https://example.com/original",
                null,
                PostUrlEntityKind.LINK,
            ),
        )

        assertEquals(
            "本文 https://not-t.co/link",
            PostTextUrlNormalizer.normalize(
                "本文 https://t.co/media https://t.co/article https://not-t.co/link",
                entities,
            ),
        )
    }

    @Test
    fun keepsTheShortUrlWhenTheDestinationIsMissingOrUnsafe() {
        val entities = listOf(
            PostUrlEntity(
                "https://t.co/docs",
                "javascript:alert(1)",
                null,
                PostUrlEntityKind.LINK,
            ),
        )

        assertEquals(
            "https://t.co/docs",
            PostTextUrlNormalizer.normalize("https://t.co/docs", entities),
        )
    }

    @Test
    fun decodesSafeHtmlReferencesExactlyOnceWithoutUrlEntities() {
        assertEquals(
            "A < B > C & D < E > F \"Q\" 'x' &unknown; &lt;",
            PostTextUrlNormalizer.normalize(
                "A &lt; B &gt; C &amp; D &#60; E &#x3E; F &quot;Q&quot; &apos;x&apos; " +
                    "&unknown; &amp;lt;",
                emptyList(),
            ),
        )
    }

    @Test
    fun leavesInvalidNumericReferencesUnchanged() {
        assertEquals(
            "&#0; &#xD800; &#99999999;",
            PostTextUrlNormalizer.normalize("&#0; &#xD800; &#99999999;", emptyList()),
        )
    }
}
